File: tickets.py
import re
import discord
from discord.ext import commands
from discord.ui import View, Button, Modal, InputText
import random
from roles import is_admin, is_staff, is_helper, is_restricted
from utils import bot_can_manage_channels, generate_ticket_transcript
from config import DEFAULT_HELPER_SLOTS, DEFAULT_POINT_VALUES

# Active tickets
active_tickets = {}

# ---------------- BOSS SELECTION ----------------
DAILY_4MAN_BOSSES = [
    "UltraEzrajal", "UltraWarden", "UltraEngineer",
    "UltraTyndarius", "UltraDage", "UltraIara", "UltraKala"
]

DAILY_7MAN_BOSSES = {
    "Astralshrine": ["Astralshrine"],
    "KathoolDepths": ["KathoolDepths"],
    "Originul": ["VoidFlibbi", "VoidNightbane", "VoidXyfrag"],
    "ApexAzalith": ["ApexAzalith"],
    "Lich Lord/Beast/Deimos": ["LichLord"],
    "Lavarockshore": ["Lavarockshore"]
}

WEEKLY_ULTRA_BOSSES = [
    "UltraDarkon", "ChampionDrakath", "UltraDage", "UltraNulgath", "UltraDrago"
]

CATEGORY_CHANNEL_PREFIX = {
    "UltraSpeaker Express": "UltraSpeaker",
    "Ultra Gramiel Express": "UltraGramiel",
    "GrimChallenge Express": "GrimChallenge",
    "Daily Temple Express": "TempleShrine",
    "Daily 4-Man Express": "4Man",
    "Daily 7-Man Express": "7Man",
    "Weekly Ultra Express": "Weekly"
}

# ---------------- PANEL VIEW ----------------
class TicketPanelView(View):
    def __init__(self, categories):
        super().__init__(timeout=None)
        for cat in categories[:25]:
            name = re.sub(r"^\d+\.\s*", "", cat["name"])
            custom_id = f"open_ticket::{name}"
            self.add_item(Button(
                label=name,
                style=discord.ButtonStyle.secondary,
                custom_id=custom_id,
                emoji="<:URE:1429522388395233331>"
            ))

# ---------------- TICKET MODAL ----------------
class TicketModal(Modal):
    def __init__(self, category_name):
        super().__init__(title=f"{category_name} Ticket")
        self.category_name = category_name
        self.add_item(InputText(label="In-game name?", required=True))
        self.add_item(InputText(label="Any concerns?", required=False))

    async def callback(self, interaction: discord.Interaction):
        in_game = self.children[0].value
        concerns = self.children[1].value or "—"

        # Random number for /join commands
        number = random.randint(1000, 99999)
        prefix = CATEGORY_CHANNEL_PREFIX.get(self.category_name, "ticket")
        channel_name = f"join-{prefix}-{number}"

        # Channel permissions
        overwrites = {
            interaction.guild.default_role: discord.PermissionOverwrite(view_channel=False),
            interaction.user: discord.PermissionOverwrite(view_channel=True, send_messages=True, read_message_history=True)
        }

        ch = await interaction.guild.create_text_channel(
            name=channel_name,
            overwrites=overwrites,
            reason=f"Ticket for {interaction.user}"
        )

        # Ticket embed
        embed = discord.Embed(
            title=f"{self.category_name} Ticket",
            description=f"Requester: {interaction.user.mention}\n**In-game name:** {in_game}\n**Concerns:** {concerns}",
            color=0x5865F2
        )

        # Add bosses and /join numbers
        if self.category_name == "Daily 4-Man Express":
            for boss in DAILY_4MAN_BOSSES:
                embed.add_field(name=boss, value=f"/join {boss}-{number}", inline=False)
        elif self.category_name == "Daily 7-Man Express":
            for boss, cmds in DAILY_7MAN_BOSSES.items():
                for cmd in cmds:
                    embed.add_field(name=boss, value=f"/join {cmd}-{number}", inline=False)
        elif self.category_name == "Weekly Ultra Express":
            for boss in WEEKLY_ULTRA_BOSSES:
                embed.add_field(name=boss, value=f"/join {boss}-{number}", inline=False)

        # Create ticket buttons view
        view = TicketActionView(interaction.user.id)

        await ch.send(embed=embed, view=view)

        # Store ticket info
        slots = DEFAULT_HELPER_SLOTS.get(self.category_name, 1)
        points = DEFAULT_POINT_VALUES.get(self.category_name, 5)
        active_tickets[ch.id] = {
            "category": self.category_name,
            "requestor": interaction.user.id,
            "helpers": [None] * slots,
            "points": points,
            "channel_id": ch.id,
            "random_number": number,
            "proof_submitted": False
        }

        await interaction.response.send_message(f"✅ Ticket created: {ch.mention}", ephemeral=True)

# ---------------- TICKET BUTTONS VIEW ----------------
class TicketActionView(View):
    def __init__(self, requestor_id):
        super().__init__(timeout=None)
        self.requestor_id = requestor_id
        self.add_item(JoinButton())
        self.add_item(SubmitProofButton(requestor_id))
        self.add_item(CloseTicketButton(requestor_id, disabled=True))  # starts disabled

class JoinButton(Button):
    def __init__(self):
        super().__init__(label="Join Ticket", style=discord.ButtonStyle.green, custom_id="join_ticket")

    async def callback(self, interaction: discord.Interaction):
        ticket_info = active_tickets.get(interaction.channel.id)
        if not ticket_info:
            await interaction.response.send_message("No active ticket found.", ephemeral=True)
            return
        if interaction.user.id == ticket_info["requestor"]:
            await interaction.response.send_message("You cannot join your own ticket.", ephemeral=True)
            return
        if interaction.user.id in [h for h in ticket_info["helpers"] if h]:
            await interaction.response.send_message("You are already helping this ticket.", ephemeral=True)
            return

        # Add helper to first empty slot
        for i in range(len(ticket_info["helpers"])):
            if ticket_info["helpers"][i] is None:
                ticket_info["helpers"][i] = interaction.user.id
                break

        try:
            await interaction.channel.set_permissions(interaction.user, view_channel=True, send_messages=True)
        except Exception:
            pass

        await interaction.response.send_message("✅ You joined the ticket.", ephemeral=True)

class SubmitProofButton(Button):
    def __init__(self, requestor_id):
        super().__init__(label="Submit Proof", style=discord.ButtonStyle.blurple, custom_id="submit_proof")
        self.requestor_id = requestor_id

    async def callback(self, interaction: discord.Interaction):
        if interaction.user.id != self.requestor_id:
            await interaction.response.send_message("Only the requestor can submit proof.", ephemeral=True)
            return

        ticket_info = active_tickets.get(interaction.channel.id)
        if ticket_info["proof_submitted"]:
            await interaction.response.send_message("Proof already submitted.", ephemeral=True)
            return

        await interaction.response.send_modal(ProofModal(interaction.channel.id))

# ---------------- CLOSE TICKET ----------------
class CloseTicketButton(Button):
    def __init__(self, requestor_id, disabled=True):  # starts disabled
        super().__init__(label="Close Ticket", style=discord.ButtonStyle.red, custom_id="close_ticket", disabled=disabled)
        self.requestor_id = requestor_id

    async def callback(self, interaction: discord.Interaction):
        ticket_info = active_tickets.get(interaction.channel.id)
        if not ticket_info:
            await interaction.response.send_message("No active ticket found.", ephemeral=True)
            return

        if interaction.user.id == ticket_info["requestor"] and not ticket_info.get("proof_submitted", False):
            await interaction.response.send_message(
                "You must submit proof before closing the ticket.",
                ephemeral=True
            )
            return

        if not (interaction.user.id == ticket_info["requestor"] or is_staff(interaction.user) or is_admin(interaction.user)):
            await interaction.response.send_message("Only staff/admin or requestor can close this ticket.", ephemeral=True)
            return

        # Remove helpers/requestor permissions
        for user_id in ticket_info["helpers"] + [ticket_info["requestor"]]:
            if user_id:
                member = interaction.guild.get_member(user_id)
                if member:
                    try:
                        await interaction.channel.set_permissions(member, view_channel=False, send_messages=False)
                    except Exception:
                        pass

        # Reward helpers
        points = ticket_info.get("points", 5)
        for helper_id in ticket_info["helpers"]:
            if helper_id:
                current = await db.get_points(helper_id)
                await db.set_points(helper_id, current + points)

# Generate transcript
transcript_channel = interaction.guild.get_channel(1357314848253542570)
if transcript_channel:
    await generate_ticket_transcript(
        ticket_info, 
        rewarded=True, 
        closer_id=interaction.user.id, 
        destination=transcript_channel
    )

# Replace embed with closed ticket embed + Delete button
embed = discord.Embed(
    title=f"{ticket_info['category']} Ticket (Closed)",
    description="Ticket closed. Only staff/admin can delete this channel.",
    color=0x5865F2
)
view = DeleteTicketView(interaction.channel.id)
await interaction.channel.send(embed=embed, view=view)

# Remove ticket from active list
active_tickets.pop(interaction.channel.id, None)

# ---------------- DELETE BUTTON ----------------
class DeleteTicketView(View):
    def __init__(self, channel_id):
        super().__init__(timeout=None)
        self.channel_id = channel_id
        self.add_item(DeleteTicketButton())

class DeleteTicketButton(Button):
    def __init__(self):
        super().__init__(label="Delete Ticket", style=discord.ButtonStyle.danger, custom_id="delete_ticket")

    async def callback(self, interaction: discord.Interaction):
        if not (is_staff(interaction.user) or is_admin(interaction.user)):
            await interaction.response.send_message("Only staff/admin can delete this channel.", ephemeral=True)
            return
        try:
            await interaction.channel.delete(reason=f"Ticket deleted by {interaction.user}")
        except Exception:
            pass

# ---------------- PROOF MODAL ----------------
class ProofModal(Modal):
    def __init__(self, channel_id):
        super().__init__(title="Submit Proof")
        self.channel_id = channel_id
        self.add_item(InputText(label="Proof URL or Description", required=False))
        self.add_item(InputText(label="Description (optional)", required=False))

    async def callback(self, interaction: discord.Interaction):
        ticket_info = active_tickets.get(self.channel_id)
        if not ticket_info:
            await interaction.response.send_message("Ticket not found.", ephemeral=True)
            return

        ticket_info["proof_submitted"] = True

        proof_url = self.children[0].value or None
        description = self.children[1].value or "—"
        ticket_info["proof"] = proof_url or description

        # Send proof to fixed channel
        proof_channel = interaction.guild.get_channel(1357332638838558862)
        if proof_channel:
            if proof_url:
                embed = discord.Embed(title="Submitted Proof", description=description, color=0x5865F2)
                embed.set_image(url=proof_url)
                await proof_channel.send(content=f"**Proof submitted by:** {interaction.user.mention}", embed=embed)
            else:
                await proof_channel.send(f"**Proof submitted by:** {interaction.user.mention}\n**Description:** {description}")

        # Reactivate Close button
        for child in interaction.message.components[0].children:
            if child.custom_id == "close_ticket":
                child.disabled = False
        await interaction.message.edit(view=interaction.message.components)

        await interaction.response.send_message(
            "✅ Proof submitted successfully. You can now close the ticket.",
            ephemeral=True
        )

# ---------------- TICKET COG ----------------
class TicketModule(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.slash_command(name="panel", description="Deploy ticket panel (staff/admin only)")
    async def panel(self, ctx: discord.ApplicationContext):
        if not (is_staff(ctx.user) or is_admin(ctx.user)):
            await ctx.respond("You don't have permission to deploy ticket panel.", ephemeral=True)
            return

        if not ctx.guild.me.guild_permissions.manage_channels:
            await ctx.respond("I need Manage Channels permission to create ticket channels.", ephemeral=True)
            return

        categories = [
            {"name": "UltraSpeaker Express"},
            {"name": "Ultra Gramiel Express"},
            {"name": "GrimChallenge Express"},
            {"name": "Daily Temple Express"},
            {"name": "Daily 4-Man Express"},
            {"name": "Daily 7-Man Express"},
            {"name": "Weekly Ultra Express"}
        ]
        view = TicketPanelView(categories)
        embed = discord.Embed(
            title="🎮 IN-GAME ASSISTANCE 🎮",
            description="Select a service below to create a help ticket.",
            color=0x5865F2
        )
        await ctx.respond(embed=embed, view=view)

@commands.slash_command(name="ticket_kick", description="Remove a helper from the ticket (staff/admin only)")
async def ticket_kick(
    self,
    ctx: discord.ApplicationContext,
    member: discord.Option(discord.Member, "Select the helper to remove")
):
    # Check permissions
    if not (is_staff(ctx.user) or is_admin(ctx.user)):
        await ctx.respond("Only staff/admin can remove helpers.", ephemeral=True)
        return

    ticket_info = active_tickets.get(ctx.channel.id)
    if not ticket_info:
        await ctx.respond("This channel is not an active ticket.", ephemeral=True)
        return

    if member.id not in ticket_info["helpers"]:
        await ctx.respond(f"{member.mention} is not a helper in this ticket.", ephemeral=True)
        return

    # Remove from helpers list
    ticket_info["helpers"] = [h if h != member.id else None for h in ticket_info["helpers"]]

    # Revoke permissions
    try:
        await ctx.channel.set_permissions(member, overwrite=None)
    except Exception:
        pass

    # Update ticket embed
    for msg in await ctx.channel.history(limit=20).flatten():  # look for ticket embed
        if msg.embeds:
            embed = msg.embeds[0]
            if "Helpers:" in embed.description:
                helpers = [h for h in ticket_info["helpers"] if h]
                helpers_text = ", ".join(f"<@{h}>" for h in helpers) if helpers else "None"
                lines = embed.description.splitlines()
                for i, line in enumerate(lines):
                    if line.startswith("Helpers:"):
                        lines[i] = f"Helpers: {helpers_text}"
                embed.description = "\n".join(lines)
                await msg.edit(embed=embed)
                break

    await ctx.respond(f"✅ {member.mention} has been removed from this ticket.", ephemeral=True)


def setup(bot):
    bot.add_cog(TicketModule(bot))

File: tickets.js
// tickets.js - Ticket System with Modal Select Menus

import {
  ActionRowBuilder,
  ButtonBuilder,
  ButtonStyle,
  EmbedBuilder,
  ModalBuilder,
  TextInputBuilder,
  TextInputStyle,
  StringSelectMenuBuilder,
  StringSelectMenuOptionBuilder,
  ChannelType,
  AttachmentBuilder
} from 'discord.js';
import { CONFIG } from './config.js';
import {
  isAdminOrStaff,
  generateRoomNumber,
  generateJoinCommands,
  createTicketEmbed,
  generateTranscript,
  createTicketPermissions
} from './utils.js';

// Create ticket panel buttons
export function createTicketPanelButtons() {
  const rows = [];
  let currentRow = new ActionRowBuilder();
  let buttonCount = 0;

  CONFIG.CATEGORIES.forEach((category, index) => {
    const label = category.replace(' Express', '');
    
    const button = new ButtonBuilder()
      .setCustomId(`ticket_open::${category}`)
      .setLabel(label)
      .setStyle(ButtonStyle.Secondary)
      .setEmoji(CONFIG.CUSTOM_EMOJI);

    currentRow.addComponents(button);
    buttonCount++;

    // Discord allows max 5 buttons per row
    if (buttonCount === 4 || index === CONFIG.CATEGORIES.length - 1) {
      rows.push(currentRow);
      currentRow = new ActionRowBuilder();
      buttonCount = 0;
    }
  });

  return rows;
}

// Create modal for categories WITH boss selection
export function createBossServerModal(category) {
  const modal = new ModalBuilder()
    .setCustomId(`ticket_modal::${category}`)
    .setTitle(category);

  // Boss selection (using TextInput as workaround - will be converted to select in interaction)
  const bossInput = new TextInputBuilder()
    .setCustomId('bosses')
    .setLabel('Select Bosses (see options below)')
    .setStyle(TextInputStyle.Short)
    .setPlaceholder('Boss selection will appear as dropdown')
    .setRequired(true);

  // Server input
  const serverInput = new TextInputBuilder()
    .setCustomId('server')
    .setLabel('Server (see options below)')
    .setStyle(TextInputStyle.Short)
    .setPlaceholder('Server selection will appear as dropdown')
    .setRequired(true);

  // In-game name
  const nameInput = new TextInputBuilder()
    .setCustomId('in_game_name')
    .setLabel('In-game name?')
    .setStyle(TextInputStyle.Short)
    .setPlaceholder('Enter your in-game name')
    .setRequired(true)
    .setMaxLength(100);

  // Concerns
  const concernsInput = new TextInputBuilder()
    .setCustomId('concerns')
    .setLabel('Any concerns?')
    .setStyle(TextInputStyle.Paragraph)
    .setPlaceholder('Optional: Any special requests or concerns')
    .setRequired(false)
    .setMaxLength(1000);

  modal.addComponents(
    new ActionRowBuilder().addComponents(bossInput),
    new ActionRowBuilder().addComponents(serverInput),
    new ActionRowBuilder().addComponents(nameInput),
    new ActionRowBuilder().addComponents(concernsInput)
  );

  return modal;
}

// Create modal for categories WITHOUT boss selection
export function createServerOnlyModal(category) {
  const modal = new ModalBuilder()
    .setCustomId(`ticket_modal_simple::${category}`)
    .setTitle(category);

  // Server input
  const serverInput = new TextInputBuilder()
    .setCustomId('server')
    .setLabel('Server (see options below)')
    .setStyle(TextInputStyle.Short)
    .setPlaceholder('Server selection will appear as dropdown')
    .setRequired(true);

  // In-game name
  const nameInput = new TextInputBuilder()
    .setCustomId('in_game_name')
    .setLabel('In-game name?')
    .setStyle(TextInputStyle.Short)
    .setPlaceholder('Enter your in-game name')
    .setRequired(true)
    .setMaxLength(100);

  // Concerns
  const concernsInput = new TextInputBuilder()
    .setCustomId('concerns')
    .setLabel('Any concerns?')
    .setStyle(TextInputStyle.Paragraph)
    .setPlaceholder('Optional: Any special requests or concerns')
    .setRequired(false)
    .setMaxLength(1000);

  modal.addComponents(
    new ActionRowBuilder().addComponents(serverInput),
    new ActionRowBuilder().addComponents(nameInput),
    new ActionRowBuilder().addComponents(concernsInput)
  );

  return modal;
}

// Create boss selection dropdown
export function createBossSelectMenu(category) {
  let bossList = [];
  
  if (category === 'Daily 4-Man Express') {
    bossList = CONFIG.DAILY_4MAN_BOSSES;
  } else if (category === 'Daily 7-Man Express') {
    bossList = CONFIG.DAILY_7MAN_BOSSES;
  } else if (category === 'Weekly Ultra Express') {
    bossList = CONFIG.WEEKLY_ULTRA_BOSSES;
  }

  const options = bossList.map(boss =>
    new StringSelectMenuOptionBuilder()
      .setLabel(boss)
      .setValue(boss)
      .setEmoji('⚔️')
  );

  const selectMenu = new StringSelectMenuBuilder()
    .setCustomId(`boss_select::${category}`)
    .setPlaceholder(`Select bosses (1-${bossList.length})`)
    .setMinValues(1)
    .setMaxValues(bossList.length)
    .addOptions(options);

  return new ActionRowBuilder().addComponents(selectMenu);
}

// Create server selection dropdown
export function createServerSelectMenu(category, selectedBosses = []) {
  const options = CONFIG.SERVERS.map(server =>
    new StringSelectMenuOptionBuilder()
      .setLabel(server)
      .setValue(server)
      .setEmoji('🌐')
  );

  const selectMenu = new StringSelectMenuBuilder()
    .setCustomId(`server_select::${category}::${JSON.stringify(selectedBosses)}`)
    .setPlaceholder('Select your server')
    .setMinValues(1)
    .setMaxValues(1)
    .addOptions(options);

  return new ActionRowBuilder().addComponents(selectMenu);
}

// Create ticket action buttons
export function createTicketActionButtons() {
  const row = new ActionRowBuilder()
    .addComponents(
      new ButtonBuilder()
        .setCustomId('ticket_join')
        .setLabel('Join Ticket')
        .setStyle(ButtonStyle.Success)
        .setEmoji('✅'),
      new ButtonBuilder()
        .setCustomId('ticket_close')
        .setLabel('Close Ticket')
        .setStyle(ButtonStyle.Danger)
        .setEmoji('🔒'),
      new ButtonBuilder()
        .setCustomId('ticket_proof')
        .setLabel('Submit Proof')
        .setStyle(ButtonStyle.Primary)
        .setEmoji('📸'),
      new ButtonBuilder()
        .setCustomId('ticket_cancel')
        .setLabel('Cancel Ticket')
        .setStyle(ButtonStyle.Secondary)
        .setEmoji('❌')
    );

  return [row];
}

// Handle ticket button click
export async function handleTicketButtonClick(interaction, db) {
  const category = interaction.customId.split('::')[1];

  // Check if category needs boss selection
  if (['Daily 4-Man Express', 'Daily 7-Man Express', 'Weekly Ultra Express'].includes(category)) {
    // Show boss selection dropdown
    const embed = new EmbedBuilder()
      .setTitle(`🎯 Select Bosses - ${category}`)
      .setDescription('Choose which bosses you need help with:')
      .setColor(CONFIG.COLORS.PRIMARY);

    const row = createBossSelectMenu(category);

    await interaction.reply({
      embeds: [embed],
      components: [row],
      ephemeral: true
    });
  } else {
    // Show server selection dropdown directly
    const embed = new EmbedBuilder()
      .setTitle(`🌍 Select Server - ${category}`)
      .setDescription('Choose which server you\'re playing on:')
      .setColor(CONFIG.COLORS.PRIMARY);

    const row = createServerSelectMenu(category, []);

    await interaction.reply({
      embeds: [embed],
      components: [row],
      ephemeral: true
    });
  }
}

// Handle boss selection
export async function handleBossSelection(interaction, db) {
  const [, category] = interaction.customId.split('::');
  const selectedBosses = interaction.values;

  // Show server selection
  const embed = new EmbedBuilder()
    .setTitle(`🌍 Select Server - ${category}`)
    .setDescription('Choose which server you\'re playing on:')
    .setColor(CONFIG.COLORS.PRIMARY);

  const row = createServerSelectMenu(category, selectedBosses);

  await interaction.update({
    embeds: [embed],
    components: [row]
  });
}

// Handle server selection
export async function handleServerSelection(interaction, db) {
  const parts = interaction.customId.split('::');
  const category = parts[1];
  const selectedBosses = parts[2] ? JSON.parse(parts[2]) : [];
  const selectedServer = interaction.values[0];

  // Show modal for in-game name and concerns
  const modal = new ModalBuilder()
    .setCustomId(`ticket_create::${category}::${JSON.stringify(selectedBosses)}::${selectedServer}`)
    .setTitle(category);

  const nameInput = new TextInputBuilder()
    .setCustomId('in_game_name')
    .setLabel('In-game name?')
    .setStyle(TextInputStyle.Short)
    .setPlaceholder('Enter your in-game name')
    .setRequired(true)
    .setMaxLength(100);

  const concernsInput = new TextInputBuilder()
    .setCustomId('concerns')
    .setLabel('Any concerns?')
    .setStyle(TextInputStyle.Paragraph)
    .setPlaceholder('Optional: Any special requests or concerns')
    .setRequired(false)
    .setMaxLength(1000);

  modal.addComponents(
    new ActionRowBuilder().addComponents(nameInput),
    new ActionRowBuilder().addComponents(concernsInput)
  );

  await interaction.showModal(modal);
}

// Handle ticket creation from modal
export async function handleTicketCreation(interaction, db) {
  await interaction.deferReply({ ephemeral: true });

  const parts = interaction.customId.split('::');
  const category = parts[1];
  const selectedBosses = parts[2] ? JSON.parse(parts[2]) : [];
  const selectedServer = parts[3];

  const inGameName = interaction.fields.getTextInputValue('in_game_name');
  const concerns = interaction.fields.getTextInputValue('concerns') || 'None';

  const guild = interaction.guild;
  const categoryChannel = guild.channels.cache.get(CONFIG.CHANNELS.TICKETS_CATEGORY);

  if (!categoryChannel) {
    return await interaction.editReply({
      content: '❌ Ticket category not configured!',
      ephemeral: true
    });
  }

  // Generate random room number
  const randomNumber = generateRoomNumber();

  // Get channel prefix
  const prefix = CONFIG.CATEGORY_METADATA[category]?.prefix || 'ticket';
  const username = interaction.user.username.toLowerCase().replace(/\s/g, '').substring(0, 20);
  const channelName = `${prefix}-${username}`;

  // Create ticket channel
  const permissions = createTicketPermissions(guild, interaction.user.id);

  try {
    const ticketChannel = await guild.channels.create({
      name: channelName,
      type: ChannelType.GuildText,
      parent: categoryChannel.id,
      permissionOverwrites: permissions
    });

    // Create ticket data
    const ticketData = {
      channel_id: ticketChannel.id,
      category: category,
      requestor_id: interaction.user.id,
      helpers: [],
      points: CONFIG.POINT_VALUES[category] || 0,
      random_number: randomNumber,
      proof_submitted: false,
      in_game_name: inGameName,
      concerns: concerns,
      selected_bosses: selectedBosses,
      selected_server: selectedServer,
      is_closed: false
    };

    // Create embed
    const embed = new EmbedBuilder(createTicketEmbed(ticketData));

    // Send ticket message
    const helperRole = CONFIG.ROLES.HELPER;
    const pingContent = `${interaction.user} ${helperRole ? `<@&${helperRole}>` : ''} ticket created!`;

    const ticketMessage = await ticketChannel.send({
      content: pingContent,
      embeds: [embed],
      components: createTicketActionButtons()
    });

    // Save to database
    ticketData.embed_message_id = ticketMessage.id;
    db.saveTicket(ticketData);

    // Generate join commands
    const joinCommands = generateJoinCommands(category, selectedBosses, randomNumber);

    // Send confirmation to user
    const confirmMessage = `✅ Ticket created: ${ticketChannel}\n\n**🎮 Your Room Number: \`${randomNumber}\`**`;
    
    await interaction.editReply({
      content: joinCommands ? `${confirmMessage}\n\n**Join Commands:**\n${joinCommands}` : confirmMessage,
      ephemeral: true
    });

  } catch (error) {
    console.error('Error creating ticket:', error);
    await interaction.editReply({
      content: `❌ Failed to create ticket: ${error.message}`,
      ephemeral: true
    });
  }
}

// Handle join ticket button
export async function handleJoinTicket(interaction, db) {
  const ticket = db.getTicket(interaction.channel.id);

  if (!ticket) {
    return await interaction.reply({
      content: '❌ No active ticket found.',
      ephemeral: true
    });
  }

  if (ticket.is_closed) {
    return await interaction.reply({
      content: '❌ This ticket is already closed.',
      ephemeral: true
    });
  }

  // Check if user is requestor
  if (interaction.user.id === ticket.requestor_id) {
    return await interaction.reply({
      content: '❌ You cannot join your own ticket.',
      ephemeral: true
    });
  }

  // Check if already a helper
  if (ticket.helpers.includes(interaction.user.id)) {
    return await interaction.reply({
      content: '❌ You are already a helper in this ticket.',
      ephemeral: true
    });
  }

  // Check helper slots
  const maxSlots = CONFIG.HELPER_SLOTS[ticket.category] || 3;
  if (ticket.helpers.length >= maxSlots) {
    return await interaction.reply({
      content: `❌ This ticket is full (${maxSlots}/${maxSlots} helpers).`,
      ephemeral: true
    });
  }

  // Check if helper is in another ticket
  const allTickets = db.getAllActiveTickets();
  for (const otherTicket of allTickets) {
    if (otherTicket.channel_id !== interaction.channel.id) {
      if (otherTicket.helpers.includes(interaction.user.id)) {
        const otherChannel = interaction.guild.channels.cache.get(otherTicket.channel_id);
        return await interaction.reply({
          content: `❌ You are already helping in another ticket: ${otherChannel || 'Unknown Channel'}\nYou can only help in ONE ticket at a time.`,
          ephemeral: true
        });
      }
    }
  }

  // Add helper
  ticket.helpers.push(interaction.user.id);
  db.saveTicket(ticket);

  // Update channel permissions
  await interaction.channel.permissionOverwrites.create(interaction.user, {
    ViewChannel: true,
    SendMessages: true
  });

  // Update ticket embed
  const updatedEmbed = new EmbedBuilder(createTicketEmbed(ticket));

  try {
    const message = await interaction.channel.messages.fetch(ticket.embed_message_id);
    await message.edit({ embeds: [updatedEmbed] });
  } catch (error) {
    console.error('Error updating ticket embed:', error);
  }

  // Send join commands
  const joinCommands = generateJoinCommands(ticket.category, ticket.selected_bosses, ticket.random_number);

  await interaction.reply({
    content: `✅ ${interaction.user} joined the ticket!\n\n**🎮 Room Number: \`${ticket.random_number}\`**\n\n**Join Commands:**\n${joinCommands || 'No commands available'}`,
    ephemeral: false
  });
}

// Handle close ticket button
export async function handleCloseTicket(interaction, db) {
  const ticket = db.getTicket(interaction.channel.id);

  if (!ticket) {
    return await interaction.reply({
      content: '❌ No active ticket found.',
      ephemeral: true
    });
  }

  if (ticket.is_closed) {
    return await interaction.reply({
      content: '❌ This ticket is already closed.',
      ephemeral: true
    });
  }

  // Check permissions (staff/admin or requestor)
  const isStaff = isAdminOrStaff(interaction.member);
  const isRequestor = interaction.user.id === ticket.requestor_id;

  if (!isStaff && !isRequestor) {
    return await interaction.reply({
      content: '❌ Only staff or the ticket creator can close tickets.',
      ephemeral: true
    });
  }

  // Check if proof was submitted
  if (!ticket.proof_submitted) {
    return await interaction.reply({
      content: '❌ Please submit proof before closing the ticket.\nUpload a screenshot showing completed quests and helper names, then click **Submit Proof**.',
      ephemeral: true
    });
  }

  await interaction.deferReply();

  // Award points to helpers
  const pointsPerHelper = ticket.points;
  let totalAwarded = 0;

  for (const helperId of ticket.helpers) {
    db.addPoints(helperId, pointsPerHelper);
    totalAwarded += pointsPerHelper;
  }

  // Mark ticket as closed
  ticket.is_closed = true;
  db.saveTicket(ticket);

  // Save to history
  db.saveTicketHistory({
    channel_id: ticket.channel_id,
    category: ticket.category,
    requestor_id: ticket.requestor_id,
    helpers: JSON.stringify(ticket.helpers),
    points_per_helper: pointsPerHelper,
    total_points_awarded: totalAwarded,
    closed_by: interaction.user.id
  });

  // Generate transcript
  const transcript = await generateTranscript(interaction.channel);

  // Send transcript to transcript channel
  const transcriptChannelId = CONFIG.CHANNELS.TRANSCRIPT;
  if (transcriptChannelId) {
    const transcriptChannel = interaction.guild.channels.cache.get(transcriptChannelId);
    if (transcriptChannel) {
      const summaryEmbed = new EmbedBuilder()
        .setTitle(`📋 Ticket Closed - ${ticket.category}`)
        .setColor(CONFIG.COLORS.SUCCESS)
        .setTimestamp()
        .addFields(
          { name: 'Requestor', value: `<@${ticket.requestor_id}>`, inline: true },
          { name: 'Helpers', value: `${ticket.helpers.length}`, inline: true },
          { name: 'Points Awarded', value: `${totalAwarded.toLocaleString()}`, inline: true },
          {
            name: 'Helper List',
            value: ticket.helpers.length > 0 ? ticket.helpers.map(id => `<@${id}>`).join('\n') : 'None',
            inline: false
          }
        )
        .setFooter({ text: `Closed by ${interaction.user.tag}` });

      const attachment = new AttachmentBuilder(
        Buffer.from(transcript, 'utf-8'),
        { name: `ticket-${ticket.channel_id}.txt` }
      );

      await transcriptChannel.send({
        embeds: [summaryEmbed],
        files: [attachment]
      });
    }
  }

  // Send close message
  const closeEmbed = new EmbedBuilder()
    .setTitle('✅ Ticket Closed')
    .setDescription(
      `**Points Awarded:** ${totalAwarded.toLocaleString()} total (${pointsPerHelper.toLocaleString()} per helper)\n` +
      `**Helpers:** ${ticket.helpers.length}\n\n` +
      'This channel will be deleted in 10 seconds...'
    )
    .setColor(CONFIG.COLORS.SUCCESS)
    .setFooter({ text: `Closed by ${interaction.user.tag}` });

  await interaction.editReply({ embeds: [closeEmbed] });

  // Delete channel after 10 seconds
  setTimeout(async () => {
    try {
      await interaction.channel.delete(`Ticket closed by ${interaction.user.tag}`);
    } catch (error) {
      console.error('Error deleting channel:', error);
    }
  }, 10000);
}

// Handle submit proof button
export async function handleSubmitProof(interaction, db) {
  const ticket = db.getTicket(interaction.channel.id);

  if (!ticket) {
    return await interaction.reply({
      content: '❌ No active ticket found.',
      ephemeral: true
    });
  }

  if (ticket.is_closed) {
    return await interaction.reply({
      content: '❌ This ticket is already closed.',
      ephemeral: true
    });
  }

  // Only requestor can submit proof
  if (interaction.user.id !== ticket.requestor_id) {
    return await interaction.reply({
      content: '❌ Only the ticket creator can submit proof.',
      ephemeral: true
    });
  }

  // Mark proof as submitted
  ticket.proof_submitted = true;
  db.saveTicket(ticket);

  await interaction.reply({
    content: '✅ Proof submitted! You can now close the ticket.',
    ephemeral: false
  });
}

// Handle cancel ticket button
export async function handleCancelTicket(interaction, db) {
  const ticket = db.getTicket(interaction.channel.id);

  if (!ticket) {
    return await interaction.reply({
      content: '❌ No active ticket found.',
      ephemeral: true
    });
  }

  // Check permissions (staff/admin or requestor)
  const isStaff = isAdminOrStaff(interaction.member);
  const isRequestor = interaction.user.id === ticket.requestor_id;

  if (!isStaff && !isRequestor) {
    return await interaction.reply({
      content: '❌ Only staff or the ticket creator can cancel tickets.',
      ephemeral: true
    });
  }

  await interaction.deferReply();

  // Mark as closed without points
  ticket.is_closed = true;
  db.saveTicket(ticket);

  const cancelEmbed = new EmbedBuilder()
    .setTitle('❌ Ticket Cancelled')
    .setDescription('This ticket was cancelled. No points were awarded.\n\nThis channel will be deleted in 5 seconds...')
    .setColor(CONFIG.COLORS.WARNING)
    .setFooter({ text: `Cancelled by ${interaction.user.tag}` });

  await interaction.editReply({ embeds: [cancelEmbed] });

  // Delete channel after 5 seconds
  setTimeout(async () => {
    try {
      await interaction.channel.delete(`Ticket cancelled by ${interaction.user.tag}`);
    } catch (error) {
      console.error('Error deleting channel:', error);
    }
  }, 5000);
}